### ISSUE TITLE
Use a bigarray directly for the internal buffer

### ISSUE THREAD

**Issue by seliopou** [inhabitedtype/angstrom#1]
This library uses the [cstruct](https://github.com/mirage/ocaml-cstruct) library for maintaining the parser's internal buffer. I made this decision early on for two reasons:

1. angstrom must maintain two regions in the buffer. The first contains input that has already been consumed by the parser and that can be overridden when provided future input. The second is contains input that has yet to be parsed. A `Cstruct.t` maintains an `off` and `len` field, where the former can be used as the boundary of the two regions. So the necessary bookkeeping was already implemented in the library; and
2. the cstruct library implements copies between `Cstruct.t`s using `memcpy` directly, which is much faster than the default implementation that OCaml provide for bigarrays.

While it was a good idea at the time, cstruct introduces a lot of unnecessary safety checks, it introduces an additional dependency on angstrom, and it introduces an interface that some libraries may be incompatible with.

Therefore, angstrom should drop the dependency on the cstruct library and use bigarrays directly. While doing this, care should be taken to port over the `memcpy` functionality from cstruct. It is very important for performance.

**Comment by rgrinberg** [comment]
@seliopou would it be possible to put the cstruct support in an additional opam package? angstorm-cstruct or something. The code living in this repo of course.

**Comment by seliopou** [comment]
Do you mean as an input representation? If so, then certainly! It would be a very small amount of code but if it means slight usability gain then it would be worth it. Just to give you an idea of what I'm talking about, currently you would feed a `Cstruct.t` into the parser state like so:

```ocaml
feed state (`Cstruct buffer)
```
Whereas after the change, you would instead do something like this:
```ocaml
feed state (`Bigarray (Cstruct.of_bigarray buffer))
```
An `angstrom.cstruct` package could expose a feed function that just takes the Cstruct variant like the feed function used to, along with all the other variants, of course.

**Comment by rgrinberg** [comment]
OK I thought cstruct covered a slightly larger surface area of the api than that. If this is the only function then it does seem frivolous.

**Comment by seliopou** [comment]
Actually, you bring up a good point that I was thinking about when I first wrote the library, but had since forgotten was a design issue. Currently, functions that perform low-level iteration such as `take`, `take_while`, etc. return strings. It may be better for them to return bigarrays instead. Or, even more radically, take an allocator function that will return a buffer of a given size, into which the parser will write the result. It may be worth implementing a general interface like this (likely in addition to what's already in place) and provide a specialized interface for cstruct.

**Comment by latticetower** [comment]
Hi, @seliopou! I've tried to work on this issue, and this is what I made now: https://github.com/latticetower/angstrom/tree/fixing_issue2
First I tried to use Bigarray directly (in different branch), but that approach lead to confusion and mess (some of this mess appeared because there were still two variables named `offset` and `off`, and I couldn't decide if one is enough). That's why I choose a simplier way - I looked at Cstruct sources and adopted code that is used by angstrom (it seems to be ok, because the license type you use and Cstruct's license are compatible). I also removed all safety checks, maybe it's bad and some of them should be kept, I don't know.
 
I removed Cstruct from dependencies, all seems to be ok, and that branch is successfully built by  travis-ci (https://travis-ci.org/latticetower/angstrom/builds/119224651).

Is it what you expected to see or I got something wrong?

> ```feed state (`Bigarray (Cstruct.of_bigarray buffer))```
I haven't changed variant name to `Bigarray` as you suggested, because this call looks illogical to me.

And what about allocator functions? Is it a part of this issue now, or just 'todo' thought?
